### DISCUSSION TITLE
L3-Performance benchmarking: Maintain main-perf dev-branch.

### DISCUSSION THREAD

**Discussion by gapisback** [undoio/l3#43] [Ideas]
We discussed the need to do controlled performance benchmarking to measure the overheads of L3-logging in comparison to other schemes such as:

1. Baseline: With no L3 or any other form of logging.
2. Logging of messages using `fprintf()` to some output file, with or w/o `fflush()`
3. Logging using the header-only [spdlog](https://github.com/gabime/spdlog) package
4. Different implement of L3-APIs where, say, `l3_init()` falls-back to `fopen()` specified file. And then `l3_log_simple()` defaults to a simple `fprintf()`
5. Extending recently-created [client-server-msgs-perf](https://github.com/undoio/l3/tree/main/use-cases/client-server-msgs-perf) performance exerciser to driver micro-benchmarking performance.

---

I think trying to maintain all scripts, drivers to do these kinds of performance-benchmarking experiments in core L3 repo-sources will become difficult.

## Proposal

1. Create a dev-branch called `main-perf-bm` which will be cut-off of `/main`.
2. All specialized perf-specific code, e.g. conditinonaly compiled code in `client-server-msgs-perf` sources will be implemented in this branch.
3. We can maintain a submodule of `spdlog` repo in this dev-branch, to simplify integration of that library with`client-server-msgs-perf` programs.
4. Any future perf-benchmarking specific sources / tools can be developed and maintained in just this one branch, which likely goes thru its own variation of CI-builds (tbd).

@gregthelaw -- Let me know if this scheme works for you and any issues you see with this scheme.

If agreeable, I can start working on this dev-branch to start building out the perf-bm tooling.

**Comment by gregthelaw** [comment]
> @gregthelaw -- Let me know if this scheme works for you and any issues you see with this scheme.

sounds good!

**Comment by gapisback** [comment]
Proposal has been accepted and moved over to [new issue 45](https://github.com/undoio/l3/issues/45).
